haskell/private/cc_wrapper.py.tpl

```diff
@@ -0,0 +1,812 @@
+                else:
+                    out.extend(["-Xlinker", ld_arg])
+            elif self._prev_ld_arg == "-rpath":
+                self._prev_ld_arg = None
+                self._handle_rpath(ld_arg, out)
+            else:
+                # This indicates a programmer error and should not happen.
+                raise RuntimeError("Unhandled _prev_ld_arg '{}'.".format(self._prev_ld_arg))
+            return True
+        elif arg.startswith("-Wl,"):
+            ld_args = arg.split(",")[1:]
+            if len(ld_args) == 2 and ld_args[0] == "-rpath":
+                self._handle_rpath(ld_args[1], out)
+                return True
+            else:
+                out.append(arg)
+                return True
+        else:
+            return False
+
+    def _handle_rpath(self, rpath, out):
+        # Filter out all RPATH flags for now and manually add the needed ones
+        # later on.
+        self.rpaths.append(rpath)
+
+    def _handle_print_file_name(self, arg, args, out):
+        if arg == "--print-file-name":
+            print_file_name = next(args)
+        elif arg.startswith("--print-file-name="):
+            print_file_name = arg[len("--print-file-name="):]
+        else:
+            return False
+
+        # Remember print-file-name action. Don't forward to allow for later
+        # manipulation.
+        self.print_file_name = print_file_name
+        self.action = Args.PRINT_FILE_NAME
+
+        return True
+
+    def _handle_compile(self, arg, args, out):
+        if arg == "-c":
+            self.action = Args.COMPILE
+            out.append(arg)
+        else:
+            return False
+
+        return True
+
+
+def load_response_files(args):
+    """Generator that loads arguments from response files.
+
+    Passes through any regular arguments.
+
+    Args:
+      args: Iterable of arguments.
+
+    Yields:
+      All arguments, with response files replaced by their contained arguments.
+
+    """
+    args = iter(args)
+    for arg in args:
+        if arg == "-install_name":
+            # macOS only: The install_name may start with an '@' character.
+            yield arg
+            yield next(args)
+        elif arg.startswith("@"):
+            with open(arg[1:], "r") as rsp:
+                for line in rsp:
+                    for rsp_arg in parse_response_line(line):
+                        yield rsp_arg
+        else:
+            yield arg
+
+
+def parse_response_line(s):
+    # GHC writes response files with quoted lines.
+    return shlex.split(s)
+
+
+def shorten_path(input_path):
+    """Shorten the given path if possible.
+
+    Applies the following transformations if they shorten the path length:
+      - Make path relative to CWD.
+      - Remove redundant up-level references.
+      - Resolve symbolic links.
+
+    Args:
+      input_path: The path to shorten.
+
+    Returns:
+      The shortened path.
+
+    """
+    exists = os.path.exists(input_path)
```

Comment:
What does it mean that a given file does not exist? Should we accept that in the first place?

Reply:
This could happen with `--library-path`. E.g. if a Haskell library has a dynamic and static output, then we'll have the static library dir (`bazel-out/.../<package>`) and the dynamic library dir (`.../_solib_XYZ/...`) listed in the package-db. But, if we're compiling in static mode, then the dynamic library won't be in the sandbox and the directory won't exist. This could also happen with Cabal libraries that have bogus `library-path` contents. GHC and GCC are lenient about this, so we should be, too. However, it's probably safer to capture this case before we even call `shorten_path`. I've changed it so that `shorten_path` requires the path to exist and is simply not called on paths that don't.